### PR TITLE
feature: add trailing commas

### PR DESCRIPTION
**Pull request by koshachy**
add trailing commas to Coding Conventions (coding-conventions.md)

### CODE REVIEW

File: pages/docs/reference/coding-conventions.md [hunk 1]
```diff
@@ -726,6 +727,213 @@ foo {
...
+    Foo, 
+    Bar, // trailing comma
+}
+```
+
+</div>
+
+#### Class parameters
+
+Place the parameters on different lines:
```
Comment (koshachy) [inline pages/docs/reference/coding-conventions.md:787]: I'm not quite sure about these repeated single lines before each code block. 
@p7nov do you have any thoughts on it?

Comment (p7nov) [reply pages/docs/reference/coding-conventions.md:787]: I don't see much value in these lines.

Comment (koshachy) [reply pages/docs/reference/coding-conventions.md]: Ok, I've eliminated them.

File: pages/docs/reference/coding-conventions.md [hunk 2]
```diff
@@ -726,6 +727,195 @@ foo {
 
 </div>
 
+### Trailing commas
+
+Use trailing commas at declaration site, including multi-line enumerations such as argument and parameter lists, `when` entries, and components of destructuring declarations.
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:732]: I'm not sure that readers are familiar with the term "trailing comma". Perhaps we should add a definition.

File: pages/docs/reference/coding-conventions.md [hunk 3]
```diff
@@ -726,6 +727,195 @@ foo {
...
+For call site trailing commas are optional.
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:733]: Isn't it optional at the declaration site as well?

Comment (p7nov) [reply pages/docs/reference/coding-conventions.md:733]: Maybe we should explain separately that it's _technically_ optional at both sites, but _as a convention_ we recommend (what we recommend :)).

File: pages/docs/reference/coding-conventions.md [hunk 4]
```diff
@@ -726,6 +727,195 @@ foo {
...
+* [Class parameters](#class-parameters)
+* [Function value parameters](#function-value-parameters)
+* [Parameters with optional type (including setters)](#parameters-with-optional-type-including-setters)
+* [Value arguments](#value-arguments)
+* [Indexing suffix](#indexing-suffix)
+* [Lambda parameters](#lambda-parameters)
+* [`when` entry](#when-entry)
+* [Collection literals (in annotations)](#collection-literals-in-annotations)
+
+> Kotlin doesn't support:
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:752]: Seems out of scope for coding conventions.

File: pages/docs/reference/coding-conventions.md [hunk 5]
```diff
@@ -726,6 +727,195 @@ foo {
...
+
+```kotlin
+val colors = listOf(
+    "red",
+    "green",
+    "blue", // trailing comma
+)
+
+enum class Classes { 
+    Foo,
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:776]: We want our examples to be more meaningful. At least by avoiding adding new `Foo`s and `Bar`s.

File: pages/docs/reference/coding-conventions.md [hunk 6]
```diff
@@ -726,6 +727,195 @@ foo {
...
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+fun foo(
+    x: Int,
+    y: Number, // trailing comma
+) {}
+
+constructor(
+    x: Comparable<Comparable<Number>>,
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:807]: Are you sure we need all these generics in generics for an example of trailing comma?

File: pages/docs/reference/functions.md [hunk 7]
```diff
@@ -43,13 +43,16 @@ Stream().read() // create instance of class Stream and call read()
 
 ### Parameters
 
-Function parameters are defined using Pascal notation, i.e. *name*: *type*. Parameters are separated using commas. 
+Function parameters are defined using Pascal notation, i.e. *name*: *type*. Parameters are separated using commas. Use a [trailing comma](coding-conventions.html#trailing-commas) when declaring function parameters.
```
Comment (p7nov) [inline pages/docs/reference/functions.md:46]: Not sure we can just recommend it in general for all functions. There are hundreds of examples without it in our own docs and Kotlin codebase :)

**Review by p7nov — CHANGES_REQUESTED** [review]

File: pages/docs/reference/coding-conventions.md [hunk 8]
```diff
@@ -726,6 +727,201 @@ foo {
...
+* It's easy to add or reorder elements: no need to add or delete the comma sign if you manipulate elements.
+* It simplifies code generation (for example, for object initializers). The last element can also have the comma sign.
+
+Using trailing commas are optional: your code will work without them. But Kotlin style guide encourages the use of trailing commas at the declaration site and leaves it optional for the call site.
+
+To enable trailing commas for the IntelliJ formatter, go to __Settings | Editor | Code Style | Kotlin__, 
+open the __Other__ tab and enable the __Use trailing comma__ option.
+
+Kotlin supports trailing commas in the following cases:
+* [Lists and enumerations](#lists-and-enumerations)
```
Comment (PetukhovVictor) [inline pages/docs/reference/coding-conventions.md:758]: The support in Lists is the same as in value arguments.

File: pages/docs/reference/coding-conventions.md [hunk 9]
```diff
@@ -726,6 +727,201 @@ foo {
...
+* It makes version-control diffs cleaner: all focus is on changed value.
+* It's easy to add or reorder elements: no need to add or delete the comma sign if you manipulate elements.
+* It simplifies code generation (for example, for object initializers). The last element can also have the comma sign.
+
+Using trailing commas are optional: your code will work without them. But Kotlin style guide encourages the use of trailing commas at the declaration site and leaves it optional for the call site.
+
+To enable trailing commas for the IntelliJ formatter, go to __Settings | Editor | Code Style | Kotlin__, 
+open the __Other__ tab and enable the __Use trailing comma__ option.
+
+Kotlin supports trailing commas in the following cases:
```
Comment (PetukhovVictor) [inline pages/docs/reference/coding-conventions.md:757]: Not all supported cases are listed here. There is also support trailing comma in type arguments, type parameters and destructuring declaration.

File: pages/docs/reference/coding-conventions.md [hunk 10]
```diff
@@ -726,6 +727,201 @@ foo {
...
+
+
+#### Collection literals (in annotations)
+
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+annotation class Anno(val x: IntArray)
+
+@Anno([1, 2, 3, 4,])
```
Comment (PetukhovVictor) [inline pages/docs/reference/coding-conventions.md:919]: I believe the elements should be placed on separate lines otherwise the trailing comma loses all its advantages.

File: pages/docs/reference/coding-conventions.md (hunk 9 above)
Comment (p7nov) [reply pages/docs/reference/coding-conventions.md:757]: Actually, this list of supported cases seems like reference-style documentation. I personally don't think that Coding Conventions page is a place for describing the language syntax in this way. 
I'd write in in more general with words with several (3-5) **typical** examples.

File: pages/docs/reference/functions.md [hunk 11]
```diff
@@ -43,13 +43,17 @@ Stream().read() // create instance of class Stream and call read()
 
 ### Parameters
 
-Function parameters are defined using Pascal notation, i.e. *name*: *type*. Parameters are separated using commas. 
+Function parameters are defined using Pascal notation, i.e. *name*: *type*. Parameters are separated using commas.
+The Kotlin code style recommends using a [trailing comma](coding-conventions.html#trailing-commas) when you declare function parameters.
```
Comment (p7nov) [inline pages/docs/reference/functions.md:47]: This is the most questionable thing as for me, so I'll ask again: can we really recommend trailing comma in function (and class) declarations **unconditionally**?
I mean, everybody declares small functions in a way like `fun sum(a: Int, b:Int): Int` - this sentence states that it's wrong now. Why? How does trailing comma make it better?

File: pages/docs/reference/coding-conventions.md (hunk 8 above)
Comment (koshachy) [reply pages/docs/reference/coding-conventions.md:758]: Is it ok if I merge them into one chapter?

I mean
`### Lists, enumerations, value arguments`

File: pages/docs/reference/functions.md (hunk 11 above)
Comment (koshachy) [reply pages/docs/reference/functions.md:47]: I don't see a contradiction here: 
Kotlin style **recommends**: recommendation != obligation

And this code line `fun sum(a: Int, b:Int): Int` is not wrong.

Comment (p7nov) [reply pages/docs/reference/functions.md:47]: I sure agree that it's not an obligation. Under "wrong" I mean "not recommended"/"contradicts with recommendations". 
The sentence says "recommend trailing comma in (any!) function parameters". So we directly say to the reader that the recommended way is different from this (my example). But this declaration is OK, isn't it?
I would expect more tailored recommendations - when is this comma useful? I really don't see a point in trailing comma in many cases (one-line declarations, for example).

File: pages/docs/reference/coding-conventions.md [hunk 12]
```diff
@@ -726,6 +727,253 @@ foo {
...
+    NORTH,
+    SOUTH,
+    WEST,
+    EAST, // trailing comma
+}
+
+fun sum(x: Any, y: Any) {}
+
+fun main() {
+    sum(
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:790]: I think `listOf()` is a more natural example for value parameters (and it doesn't require declaration and `main` :)
```
val numbers = listOf(
    "one",
    "two",
    "three", // trailing comma
)
```

File: pages/docs/reference/coding-conventions.md [hunk 13]
```diff
@@ -726,6 +727,253 @@ foo {
...
+</div>
+
+#### Class parameters
+
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+class Customer(
+    val name: String,
+    val surname: String, // trailing comma
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:806]: Nitpicking: _last name_ is preferable over _surname_ in American English.

File: pages/docs/reference/coding-conventions.md [hunk 14]
```diff
@@ -726,6 +727,253 @@ foo {
...
+
+</div>
+
+#### Indexing suffix
+
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+class Surface {
+    operator fun get(x: Int, y: Int) = 2*x + 4*y - 10
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:858]: Kotlin coding conventions (aka this page :) ) says [put spaces around binary operators (a + b).](https://kotlinlang.org/docs/reference/coding-conventions.html#horizontal-whitespace)

File: pages/docs/reference/coding-conventions.md [hunk 15]
```diff
@@ -726,6 +727,253 @@ foo {
...
+
+</div>
+
+#### Collection literals (in annotations)
+
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+annotation class ApplicableFor(val services: Array<String>)
+@ApplicableFor([
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:910]: Pls add a blank line between annotation and annotated function.

File: pages/docs/reference/coding-conventions.md [hunk 16]
```diff
@@ -726,6 +727,253 @@ foo {
...
+```
+
+</div>
+
+#### Destructuring declarations
+
+<div class="sample" markdown="1" theme="idea" mode="kotlin" data-highlight-only>
+
+```kotlin
+fun main() {
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:948]: You don't necessarily need `main` for non-runnable code blocks.

File: pages/docs/reference/functions.md (hunk 11 above)
Comment (koshachy) [reply pages/docs/reference/functions.md]: I understand the concern; thus I rewrote this part.

File: pages/docs/reference/coding-conventions.md [hunk 17]
```diff
@@ -726,6 +727,242 @@ foo {
...
+Using trailing commas has several benefits:
+
+* It makes version-control diffs cleaner: all focus is on changed value.
+* It's easy to add or reorder elements: no need to add or delete the comma sign if you manipulate elements.
+* It simplifies code generation (for example, for object initializers). The last element can also have the comma sign.
+
+Using trailing commas are optional: your code will work without them. But Kotlin style guide encourages the use of trailing commas at the declaration site and leaves it optional for the call site.
+
+To enable trailing commas for the IntelliJ formatter, go to __Settings | Editor | Code Style | Kotlin__, 
+open the __Other__ tab and enable the __Use trailing comma__ option.
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:755]:
```suggestion
open the __Other__ tab and select the __Use trailing comma__ option.
```
See [MS Guide on UI description](https://docs.microsoft.com/en-us/style-guide/procedures-instructions/describing-interactions-with-ui)

File: pages/docs/reference/coding-conventions.md (hunk 8 above)
Comment (PetukhovVictor) [reply pages/docs/reference/coding-conventions.md]: I suggest to have only enumeration chapter and `value arguments` as `Lists` is subset of `value arguments`. `listOf(a, b, c)` is just a function call.

File: pages/docs/reference/coding-conventions.md (hunk 9 above)
Comment (PetukhovVictor) [reply pages/docs/reference/coding-conventions.md:757]: OK, I just thought that here describes all of the supported cases (due to the phrase `Kotlin supports trailing commas in the following cases`). BTW it can be merged in the following way: `value and type arguments`, `value and type parameters`.

File: pages/docs/reference/coding-conventions.md [hunk 18]
```diff
@@ -726,6 +727,242 @@ foo {
...
+
+</div>
+
+Using trailing commas has several benefits:
+
+* It makes version-control diffs cleaner: all focus is on changed value.
+* It's easy to add or reorder elements: no need to add or delete the comma sign if you manipulate elements.
+* It simplifies code generation (for example, for object initializers). The last element can also have the comma sign.
+
+Using trailing commas are optional: your code will work without them. But Kotlin style guide encourages the use of trailing commas at the declaration site and leaves it optional for the call site.
```
Comment (p7nov) [inline pages/docs/reference/coding-conventions.md:752]:
```suggestion
Trailing commas are optional: your code will work without them. Kotlin style guide encourages the use of trailing commas at the declaration site and leaves it optional for the call site.
```
- With `Using`, the 'are' verb is incorrect.
- `But` doesn't seem necessary and starting a sentence with it is an arguable topic, so I'd just remove it.
- `optional` is used in two different meanings in one line. We should replace it with smth that conducts `up to you` meaning better.

File: pages/docs/reference/functions.md [hunk 19]
```diff
@@ -53,6 +53,20 @@ fun powerOf(number: Int, exponent: Int) { /*...*/ }
 ```
 </div>
 
+The Kotlin code style recommends using a [trailing comma](coding-conventions.html#trailing-commas) when you declare function parameters.
+It simplifies reordering, deleting, and adding new parameters. Using trailing commas are optional: without them your code will work as well.
```
Comment (p7nov) [inline pages/docs/reference/functions.md:57]: Do we really need to repeat this on three pages? "Optional" is obvious from the example above and the `recommends` word, benefits are described in coding conventions. The same goes for the Classes page.

File: pages/docs/reference/classes.md [hunk 20]
```diff
@@ -114,6 +114,22 @@ class Person(val firstName: String, val lastName: String, var age: Int) { /*...*
 
 </div>
 
+The Kotlin code style recommends using a [trailing comma](coding-conventions.html#trailing-commas) when you declare class properties.
+It simplifies reordering, deleting, and adding new properties. Using trailing commas are optional: without them your code will work as well.
+
+<div class="sample" markdown="1" theme="idea" data-highlight-only>
+
+```kotlin
```
Comment (p7nov) [inline pages/docs/reference/classes.md:122]: Maybe it's better to rewrite the class from the example above to show only the syntax difference?
Now there are structure differences, so it's harder to find the difference we want to show.

File: pages/docs/reference/functions.md (hunk 19 above)
Comment (koshachy) [reply pages/docs/reference/functions.md:57]: IIRC, you've mentioned above that you want to see here the benefits of using trailing commas ("when is this comma useful?"). Did I miss something?

File: pages/docs/reference/coding-conventions.md (hunk 8 above)
Comment (koshachy) [reply pages/docs/reference/coding-conventions.md]: To recap.
No more `Lists and enumerations`: there are `Enumerations` and `Value arguments`.

File: pages/docs/reference/coding-conventions.md (hunk 9 above)
Comment (koshachy) [reply pages/docs/reference/coding-conventions.md:757]: I've added `Type arguments and type parameters` and `Destructuring declarations` instead of merging

File: pages/docs/reference/functions.md [hunk 21]
```diff
@@ -53,6 +53,18 @@ fun powerOf(number: Int, exponent: Int) { /*...*/ }
 ```
 </div>
 
+You can use a [trailing comma](coding-conventions.html#trailing-commas) when you declare class properties:
```
Comment (p7nov) [inline pages/docs/reference/functions.md:56]:
```suggestion
You can use a [trailing comma](coding-conventions.html#trailing-commas) when you declare function parameters:
```

**Review by p7nov — APPROVED** [review]
LGTM now.

**Review by PetukhovVictor — APPROVED** [review]
